File: lib/models/booking.dart
import 'service_item.dart';

enum BookingStatus { pending, confirmed, completed, cancelled }

extension BookingStatusLabel on BookingStatus {
  String get label {
    switch (this) {
      case BookingStatus.pending:
        return 'Chờ xác nhận';
      case BookingStatus.confirmed:
        return 'Đã xác nhận';
      case BookingStatus.completed:
        return 'Hoàn thành';
      case BookingStatus.cancelled:
        return 'Đã huỷ';
    }
  }
}

class Booking {
  final String id;
  final String userId;
  final String salonId;
  final String salonName;
  final List<ServiceItem> services;
  final String? stylistId;
  final String stylistName; // "Bất kỳ" if stylistId is null
  final DateTime date;
  final String timeSlot; // e.g. "14:30"
  final BookingStatus status;
  final String? note;

  const Booking({
    required this.id,
    required this.userId,
    required this.salonId,
    required this.salonName,
    required this.services,
    required this.stylistId,
    required this.stylistName,
    required this.date,
    required this.timeSlot,
    required this.status,
    this.note,
  });

  int get totalPriceVnd => services.fold(0, (sum, s) => sum + s.priceVnd);

  int get totalDurationMinutes =>
      services.fold(0, (sum, s) => sum + s.durationMinutes);

  Booking copyWith({BookingStatus? status}) {
    return Booking(
      id: id,
      userId: userId,
      salonId: salonId,
      salonName: salonName,
      services: services,
      stylistId: stylistId,
      stylistName: stylistName,
      date: date,
      timeSlot: timeSlot,
      status: status ?? this.status,
      note: note,
    );
  }
}

File: backend/src/models/Booking.js
const { Schema, model } = require('mongoose');

// Services are embedded as a snapshot (id/name/price/duration at booking time)
// instead of a join table, per the NoSQL mapping in docs/database-design.md.
const bookedServiceSchema = new Schema(
  {
    serviceId: { type: Schema.Types.ObjectId, ref: 'Service', required: true },
    name: { type: String, required: true },
    priceVnd: { type: Number, required: true },
    durationMinutes: { type: Number, required: true },
  },
  { _id: false },
);

const bookingSchema = new Schema(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    salonId: { type: Schema.Types.ObjectId, ref: 'Salon', required: true },
    salonName: { type: String, required: true },
    services: { type: [bookedServiceSchema], required: true },
    stylistId: { type: Schema.Types.ObjectId, ref: 'Stylist', default: null },
    stylistName: { type: String, default: 'Bất kỳ' },
    date: { type: Date, required: true },
    timeSlot: { type: String, required: true },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'completed', 'cancelled'],
      default: 'pending',
    },
    note: { type: String, default: null },
  },
  { timestamps: true },
);

bookingSchema.methods.toPublicJSON = function toPublicJSON() {
  return {
    id: this._id.toString(),
    userId: this.userId.toString(),
    salonId: this.salonId.toString(),
    salonName: this.salonName,
    services: this.services.map((s) => ({
      id: s.serviceId.toString(),
      name: s.name,
      priceVnd: s.priceVnd,
      durationMinutes: s.durationMinutes,
    })),
    stylistId: this.stylistId ? this.stylistId.toString() : null,
    stylistName: this.stylistName,
    date: this.date.toISOString(),
    timeSlot: this.timeSlot,
    status: this.status,
    note: this.note,
  };
};

module.exports = model('Booking', bookingSchema);
